nix-store: implement --query --unsubstitutable

nrdxp opened the pull request:
Resolves #3946 by providing an efficient means of querying cache status of the passed paths.

For every path passed, it will be returned back on standard output only if it is not available in any of the configured substituters.

~~A threadpool is used to make this as quick as possible.~~

Included in the documentation is a sample command showing how users can now trivially capture any and all uncached build & runtime dependencies of a given derivation. This information is highly useful in CI environments, for example.

Only caveat is that we might hit the OS arg limit until #7437 is fixed.

#### TODO
- [x] implement `nix-store` query
- [ ] write tests

# Checklist for maintainers

<!-- Contributors: please leave this as is -->

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or bug fix: updated release notes

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-to-correctly-cache-build-time-dependencies-using-nix-haskellforall-com/22726/5

mlyxshi commented:
Are you interested in this relative issue?
https://github.com/NixOS/nix/issues/7527

nrdxp commented:
> Are you interested in this relative issue? #7527

You could use this PR to accomplish your goal by simply processing you derivations through `nix-store -qU` before sending them to `nix copy`. As Sandro rightly pointed out though, it will still result in a few redundant dependencies, since if the entire dependency closure does not exist in a single cache, Nix throws an error. To improve the situation we just have to allow Nix to continue instead of erroring in this situation, and you could then just pass a `--no-recursive` to `nix copy` and cache only those that don't exist elsewhere.

In general, my recent experience with Nix in CI has kind of changed my perspective somewhat, and I'm starting to feel like we should take a more "git-like" approach to our cli, rather than replacing the older, more verbose commands outright, we should stabalize them as much as possible and use them as the "plumbing" to the higher level commands.

nrdxp commented:
Because there isn't really a nice standalone solution for this atm, I went ahead and extracted the logic as a standalone tool for now until this is accepted:
https://github.com/divnix/nix-uncached

edolstra commented:
I think functionality like this should be added to the new CLI rather than the legacy CLI. It seems to me that what we really want here is a CLI interface to `Store::queryMissing()` (ideally with JSON output).

Also "uncached" is a vague term. What does it mean for a path to be uncached? Maybe "unsubstitutable" is better.

nrdxp commented:
> I think functionality like this should be added to the new CLI rather than the legacy CLI. It seems to me that what we really want here is a CLI interface to `Store::queryMissing()` (ideally with JSON output).

I can add it to `nix path-info` I think, fairly easily.

My first implementation did just use queryMissing, but that isn't _exactly_ what I wanted, since I don't care if the path exists locally or not, I just want to know if it exists in a substituter. If you want to include the local store as part of the equation you can add `--extra-substituters auto` to the invocation to make that explicit.

> Also "uncached" is a vague term. What does it mean for a path to be uncached? Maybe "unsubstitutable" is better.

Yeah I was thinking that myself, for now I just used the terminology that has been used elsewhere in the community so far, but I was also thinking of using unsubstitutable instead just last night, only thing is that it's quite long, although I guess with a short flag it would still be fine.

nrdxp commented:
@edolstra, I implemented the same logic for `nix path-info`, including allowing it the `--json` to add a new boolean output showing whether a path is cached or not.

However there are a few downsides to the `nix path-info` implementation compared to the `nix-store` one so far. In order to explain that, understand that my main usecase for this is querying information about the paths before they have been built, to decide if they need to be built, etc. I the paths I need from a flake using `nix eval` which produces a special json we then want to manipulate based on information we are going to query, like the cache status.

Importantly, I only want to run evaluation once, and only once, for a given CI run, so passing flake refs around isn't desirable. Instead we query information on the derivations directly. This just doesn't work well with `nix path-info`.

With the `nix-store` implementation it works perfectly, I can just query the derivation with `nix-store -qR --include-outputs` and then filter that through my new `nix-store -qU` and I've got the information I need, minus some additional post-processing which can be easily done with `jq` and `nix show-derivation` (to remove locally built derivations from the equation, etc). 

Specifically these are the quircks I've encountered with the `nix path-info` implementation so far:
* as already mentioned, there is no equivalent to the `--include-outputs` flag, so its either querying buildtime _or_ runtime deps, not both.
* If I pass a derivation (which is all I have), then the command will not return information about it's outpath. I have to explicitly pass the outpath for that, but that is a non-starter for my usecase since I don't have them built yet (and don't want to necessarily).

For these reasons I propose we keep both implementations until both are more in-line, and as mentioned in my previous comment. As mentioned in my previous comment, we may want to consider a more "plumbing" and "porceline" approach for flakes 2.0, to keep drift between the old and new commands from becoming a big issue.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/what-would-you-like-to-see-improved-in-nix-cli-experience/24012/58

tfc commented on src/nix/path-info.cc:
```diff
@@ -87,41 +96,98 @@ struct CmdPathInfo : StorePathsCommand, MixJSON
+                    clone["substitute"] = !infos.empty();
+
+                    auto jsonList(jsonList_.lock());
+                    jsonList->emplace_back(clone);
```
in order to really emplace it at the back without unneeded copies i suggest this:
```suggestion
                    jsonList->emplace_back(std::move(clone));
```

tfc commented on src/nix-store/nix-store.cc:
```diff
@@ -364,6 +369,32 @@ static void opQuery(Strings opFlags, Strings opArgs)
+                auto path = store->followLinksToStorePath(i);
+
+                pool.enqueue(std::bind(queryPath, path));
+            }
```
Does anything speak against leaving out the `std::function` and `std::bind` by writing it like this: ?

```cpp
            auto queryPathThunk = [&](const StorePath & req) {
                return [&]{
                    SubstitutablePathInfos infos;
                    store->querySubstitutablePathInfos({{req, std::nullopt}}, infos);

                    if (infos.empty()) {
                        cout << store->printStorePath(req) << '\n';
                    }
                };
            };

            for (auto & i : opArgs) {
                pool.enqueue(queryPathThunk(store->followLinksToStorePath(i)));
            }
```

This would remove some typing information that is both unneeded and fixed otherwise.

tfc commented on src/nix/path-info.cc:
```diff
@@ -87,41 +96,98 @@ struct CmdPathInfo : StorePathsCommand, MixJSON
+                }
+
+                std::function<void(nlohmann::json::basic_json::value_type)> checkSubs;
+                checkSubs = [&](const nlohmann::json::basic_json::value_type & value) {
```
same suggestion with the nested lambda expression here. all the type info is relatively noisy and can be reduced.

tfc commented on src/nix/path-info.cc:
```diff
@@ -87,41 +96,98 @@ struct CmdPathInfo : StorePathsCommand, MixJSON
+                    auto clone = value;
+                    std::string storePathS = value["path"];
+                    auto info = store->queryPathInfo(store->followLinksToStorePath(storePathS));
+                    auto storePath = info->path;
```
it seems like some `const ... &` could help the reader understand what really needs to be a (mutable) copy and what not, as well as reduce some redundant copying.

tfc commented on src/nix/path-info.cc:
```diff
@@ -87,41 +96,98 @@ struct CmdPathInfo : StorePathsCommand, MixJSON
+            Sync<std::set<std::string>> results_;
+
+            std::function<void(StorePath)> printPath;
+            printPath = [&](const StorePath & storePath) {
```
Here, the parameter type from the `std::function` and the actual assignment differ by having and not having `const ... &` qualifiers. i suggest using the same nested lambda expression approach here, too.

tfc commented on src/nix/path-info.cc:
```diff
@@ -87,41 +96,98 @@ struct CmdPathInfo : StorePathsCommand, MixJSON
 
-                std::cout << std::endl;
+                auto results(results_.lock());
+                results->emplace(storePathS);
```
using `std::move` will help avoiding a needless copy.

tfc commented on src/nix/path-info.cc:
```diff
@@ -87,41 +96,98 @@ struct CmdPathInfo : StorePathsCommand, MixJSON
             }
+            pool.process();
 
+            auto results(results_.lock());
+            for (auto & str : *results) {
+                std::cout << std::endl << str;
+            }
```
`std::endl` means to emit a new line character *and* flush the terminal. But this code suggests that you mean to separate all strings by newlines and then flush only once at the end, so i suggest this:

```suggestion
            auto results(results_.lock());
            for (const auto & str : *results_.lock()) {
                std::cout << '\n' << str;
            }
            std::cout << std::endl;
```

nrdxp commented:
Thanks for the review, I'm moving to draft for now as the implementation needs work. After some benchmarks, the threadpool didn't actually help anything and is about the same as using a single thread, even when setting the threadpool limit to something super high like 1000.

Its possible that an async implementation may be superior or there is some bottleneck that needs to be tracked down. Until that is worked out, I'll leave in draft.

nrdxp commented:
Found the performance bottleneck. Turns out we don't need a threadpool at all but can get really great performance from the `Store::querySubstitutablePaths` function itself. Also implemented the same logic for `nix path-info` in a separate PR: #7587 as requested by @edolstra 

I split the two up so that one does not block merging of the other if we decide not to keep both.

edolstra commented on src/nix-store/nix-store.cc:
```diff
@@ -364,6 +365,39 @@ static void opQuery(Strings opFlags, Strings opArgs)
+
+            StorePathSet storePaths;
+            for (auto const & str : opArgs) {
+               storePaths.emplace(store->followLinksToStorePath(str));
```
The indentation here and below is wrong (should be 4 characters per level).

edolstra commented on src/nix-store/nix-store.cc:
```diff
@@ -364,6 +365,39 @@ static void opQuery(Strings opFlags, Strings opArgs)
+            }
+
+            for ( auto & path : storePaths) {
+                if (substitutablePaths.find(path) == substitutablePaths.end())
```
```suggestion
                if (!substitutablePaths.count(path))
```

ncfavier replied:
This does not short-circuit

edolstra replied:
In what way? On an `std::map` it should be equally efficient.

But since we're fine with C++20 now, you can also use `substitutablePaths.contains(path)`.

nrdxp commented:
closing in favour of the aforementioned PR

tomberek commented:
Miss click

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2023-01-20-nix-team-meeting-minutes-25/25432/1
